lib/spack/spack/cmd/buildcache.py

```diff
@@ -272,6 +380,163 @@ def getkeys(args):
+
+    # Next see if there are any configured binary mirrors
+    configured_mirrors = spack.config.get('mirrors', scope=args.scope)
+
+    if args.mirror_url:
+        configured_mirrors = {'additionalMirrorUrl': args.mirror_url}
+
+    if not configured_mirrors:
+        tty.msg('No mirrors provided, exiting.')
+        sys.exit(0)
+
+    sys.exit(bindist.check_specs_against_mirrors(
+        configured_mirrors, specs, args.output_file, args.rebuild_on_error))
+
+
+def get_tarball(args):
+    """Download buildcache entry from a remote mirror to local folder.  This
+    command uses the process exit code to indicate its result, specifically,
+    a non-zero exit code indicates that the command failed to download at
+    least one of the required buildcache components.  Normally, just the
+    tarball and .spec.yaml files are required, but if the --require-cdashid
+    argument was provided, then a .cdashid file is also required."""
+    if not args.spec and not args.spec_yaml:
+        tty.msg('No specs provided, exiting.')
+        sys.exit(0)
+
+    if not args.path:
+        tty.msg('No download path provided, exiting')
+        sys.exit(0)
+
+    spec = get_concrete_spec(args)
+
+    tarfile_name = bindist.tarball_name(spec, '.spack')
+    tarball_dir_name = bindist.tarball_directory_name(spec)
+    tarball_path_name = os.path.join(tarball_dir_name, tarfile_name)
+    local_tarball_path = os.path.join(args.path, tarball_dir_name)
+
+    files_to_fetch = [
+        {
+            'url': tarball_path_name,
+            'path': local_tarball_path,
+            'required': True,
+        }, {
+            'url': bindist.tarball_name(spec, '.spec.yaml'),
+            'path': args.path,
+            'required': True,
+        }, {
+            'url': bindist.tarball_name(spec, '.cdashid'),
+            'path': args.path,
+            'required': args.require_cdashid,
+        },
+    ]
+
+    result = bindist.download_buildcache_entry(files_to_fetch)
+
+    if result:
+        sys.exit(0)
+
+    sys.exit(1)
+
+
+def get_concrete_spec(args):
+    spec_str = args.spec
+    spec_yaml_path = args.spec_yaml
+
+    if not spec_str and not spec_yaml_path:
+        tty.msg('Must provide either spec string or path to ' +
+                'yaml to concretize spec')
+        sys.exit(1)
+
+    if spec_str:
+        try:
+            spec = Spec(spec_str)
+            spec.concretize()
+        except SpecError as spec_error:
+            tty.error('Unable to concrectize spec {0}'.format(args.spec))
+            tty.debug(spec_error)
+            sys.exit(1)
+
+        return spec
+
+    with open(spec_yaml_path, 'r') as fd:
+        return Spec.from_yaml(fd.read())
+
+
+def get_buildcache_name(args):
+    """Get name (prefix) of buildcache entries for this spec"""
+    spec = get_concrete_spec(args)
+    buildcache_name = bindist.tarball_name(spec, '')
+
+    print('{0}'.format(buildcache_name))
+
+    sys.exit(0)
+
+
+def save_dependent_spec_yaml(args):
```

Comment:
IMO this would be a good utility function, not just something available to `build_cache`. Also, I think it would be better if it read in a `.yaml` spec file (so that you didn't have to specify the root), and extracted the dependency yaml from that.

Reply:
In that case, do you think it should live in `lib/spack/spack/spec.py`?  Should it be a module-level method or a static member of `Spec`?  And regardless of that decision, what do you think of arguments like:

```python
def save_dependency_spec_yamls(root_spec_as_yaml, output_directory, dependencies=None)
```

And the caller would pass `dependencies` (a list of spec names) if they want filter out which dependencies get written, otherwise, it would write out a `.spec.yaml` for every dependency.  This would be convenient for my build jobs, which would rather just call one spack command and get all the full specs it cares about written to disk at once.

Also note that my build jobs don't have their hands on the root spec yaml file, and need to concretize the abstract root spec themselves, and so it might be nice if this method could take the root spec as a yaml object rather than insisting on reading it from a file.

But let me know what you think.

Reply:
> In that case, do you think it should live in lib/spack/spack/spec.py

That sounds good

> Should it be a module-level method or a static member of Spec

Either is fine

> what do you think of arguments like:

Those arguments look good

> note that my build jobs don't have their hands on the root spec yaml file, and need to concretize the abstract root spec themselves, and so it might be nice if this method could take the root spec as a yaml object rather than insisting on reading it from a file.

That also works